mplex/src/lib.rs

```diff
@@ -138,18 +197,45 @@ where C: AsyncRead + AsyncWrite,
     }
 
     if let Some((offset, out)) = inner.buffer.iter().enumerate().filter_map(|(n, v)| filter(v).map(|v| (n, v))).next() {
+        // The buffer was full and no longer is, so let's notify everything.
+        if inner.buffer.len() == inner.config.max_buffer_len {
+            for task in inner.to_notify.drain() {
+                task.1.notify();
+            }
+        }
+
         inner.buffer.remove(offset);
         return Ok(Async::Ready(Some(out)));
     }
 
+    // TODO: replace with another system
+    static NEXT_TASK_ID: AtomicUsize = AtomicUsize::new(0);
+    task_local!{
+        static TASK_ID: usize = NEXT_TASK_ID.fetch_add(1, Ordering::Relaxed)
+    }
+
     loop {
+        // Check if we reached max buffer length first.
+        if inner.buffer.len() == inner.config.max_buffer_len {
```

Comment:
Why not a `>=`? Can we be sure to always reach the `max_buffer_len` exactly?

Reply:
I'll add a `debug_assert!` instead, so that we detect logic errors in the code.

Reply:
if you think that satisfied it. We had a `>=` before...